parser/src/main/kotlin/com/linkedin/dex/parser/AnnotationUtils.kt

```diff
@@ -33,49 +27,72 @@ fun DexFile.getAnnotationsDirectory(classDefItem: ClassDefItem): AnnotationsDire
         return emptyList()
     }
 
     val classAnnotationSetItem = AnnotationSetItem.create(byteBuffer, directory.classAnnotationsOff)
-    return classAnnotationSetItem.toDescriptorList(this)
+
+    val values = classAnnotationSetItem.entries.map { AnnotationItem.create(byteBuffer, it.annotationOff) }
+            .map {
+                val name = formatDescriptor(ParseUtils.parseDescriptor(byteBuffer,
+                    typeIds[it.encodedAnnotation.typeIdx], stringIds))
+                val encodedAnnotationValues = it.encodedAnnotation.elements
+                val values = HashMap<String, DecodedValue>()
+                for (encodedAnnotationValue in encodedAnnotationValues) {
+                    val value = DecodedValue.createFromDecodedValue(this, encodedAnnotationValue.value)
+                    byteBuffer.position(this.stringIds[encodedAnnotationValue.nameIdx].stringDataOff)
+                    val name = ParseUtils.parseStringBytes(byteBuffer)
+
+                    values.put(name, value)
+                }
+
+                TestAnnotation(name, values)
+            }
+
+    return values
 }
 
 /**
- * @return A list of descriptors containing all the method-level annotations on the given [MethodIdItem]
+ * @return A list of annotation objects for all the method-level annotations
  */
-fun DexFile.getMethodAnnotationDescriptors(methodId: MethodIdItem,
-                                           annotationsDirectory: AnnotationsDirectoryItem?): List<String> {
-    // find the annotated method matching the given one, otherwise return an empty list
+fun DexFile.getMethodAnnotationValues(methodId: MethodIdItem, annotationsDirectory: AnnotationsDirectoryItem?): List<TestAnnotation> {
     val methodAnnotations = annotationsDirectory?.methodAnnotations ?: emptyArray<MethodAnnotation>()
-    return methodAnnotations.filter { methodIds[it.methodIdx] == methodId }
-            .flatMap { (_, annotationsOff) ->
-                val methodAnnotationSetItem = AnnotationSetItem.create(byteBuffer, annotationsOff)
-                methodAnnotationSetItem.toDescriptorList(this)
+    val annotationSets = methodAnnotations.filter { methodIds[it.methodIdx] == methodId }
+            .map { (_, annotationsOff) ->
+                AnnotationSetItem.create(byteBuffer, annotationsOff)
             }
-}
 
-/**
- * Merge the given [methodAnnotationDescriptors] and [classAnnotationDescriptors] lists and format them as
- * human-readable names
- */
-fun getAnnotationNames(methodAnnotationDescriptors: List<String>,
-                       classAnnotationDescriptors: List<String>): List<String> {
-    return methodAnnotationDescriptors
-            .plus(classAnnotationDescriptors)
-            .distinct()
-            .map(::formatDescriptor)
+
+    val allAnnotations = annotationSets.map {
+        it.entries.map { AnnotationItem.create(byteBuffer, it.annotationOff) }
+            .map {
```

Comment:
Could we move the contents of this map call to a helper method to avoid the ugly multiline implicit return?